### PR TITLE
#1536 Added support for JMS exclusion strategies

### PR DESCRIPTION
**Pull request by bblue**
Fix for issue #1536. I don't think it needs more magic than this.

### CODE REVIEW

**Comment by lsmith77** [comment]
fine for me

File: Serializer/JMSSerializerAdapter.php [hunk 1]
```diff
@@ -96,6 +96,11 @@ private function convertContext(Context $context, $direction)
         if (null !== $context->getSerializeNull()) {
             $jmsContext->setSerializeNull($context->getSerializeNull());
         }
+        if (is_array($strategies = $context->getExclusionStrategies())) {
```
Comment (Ener-Getick) [inline Serializer/JMSSerializerAdapter.php]: why not just always make it an array ?

**Comment by yagobski** [comment]
Please release this commit soon ! Thanks

File: Serializer/JMSSerializerAdapter.php (hunk 1 above)
Comment (bblue) [inline Serializer/JMSSerializerAdapter.php]: Sure can. I will update the PR when I get home today. The `is_array` should move to the `setExclusionStrategies()` method instead then. I suppose the same method should throw an exception if the parameter is not an array, so it does not fail silently (as the current approach also would).

Comment (bblue) [inline Serializer/JMSSerializerAdapter.php]: On second thought `is_array ` and an exception is not needed, any invalid data will clearly make itself known when the foreach fails. Pushed a simplied version just now.

**Comment by bblue** [comment]
I am not really sure why the last commit failed the Travis test...

File: Context/Context.php [hunk 2]
```diff
@@ -257,4 +261,46 @@ public function getSerializeNull()
...
+    /**
+     * Sets the exclusion strategies.
+     *
+     * Warning: This only applies to the JMS serializer adapter.
+     *
+     * @see https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1536
+     *
+     * @param array $exclusionStrategies
+     */
+    public function setExclusionStrategies($exclusionStrategies)
```
Comment (Ener-Getick) [inline Context/Context.php]: Do we really need a setter? Isn't an adder enough?

File: Context/Context.php [hunk 3]
```diff
@@ -257,4 +261,32 @@ public function getSerializeNull()
     {
         return $this->serializeNull;
     }
+
+    /**
+     * Gets an array of exclusion strategies.
+     *
+     * Warning: This only applies to the JMS serializer adapter.
+     *
+     * @see https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1536
```
Comment (Ener-Getick) [inline Context/Context.php]: I don't think we need to link this issue.

File: Context/Context.php [hunk 4]
```diff
@@ -257,4 +261,32 @@ public function getSerializeNull()
...
+    /**
+     * Adds an exclusion strategy.
+     *
+     * Warning: This only applies to the JMS serializer adapter.
+     *
+     * @see https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1536
+     *
+     * @param $exclusionStrategy
+     */
+    public function addExclusionStrategy($exclusionStrategy)
```
Comment (Ener-Getick) [inline Context/Context.php]: can't we type hint on an interface?

**Comment by Ener-Getick** [comment]
LGTM :+1: btw could you add some tests to avoid regressions?

File: Tests/Context/ContextTest.php [hunk 5]
```diff
@@ -98,4 +98,20 @@ public function testSerializeNull()
...
+    {
+        $strategy1 = $this->createMock('JMS\Serializer\Exclusion\ExclusionStrategyInterface');
+        $strategy2 = $this->createMock('JMS\Serializer\Exclusion\ExclusionStrategyInterface');
+
+        $this->context->addExclusionStrategy($strategy1);
+        $this->context->addExclusionStrategy($strategy2);
+
+        $strategies = $this->context->getExclusionStrategies();
+
+        $this->assertEquals(true, is_array($strategies));
```
Comment (Ener-Getick) [inline Tests/Context/ContextTest.php]: This looks useless

File: Tests/Context/ContextTest.php [hunk 6]
```diff
@@ -98,4 +98,20 @@ public function testSerializeNull()
...
+        $this->assertEquals(true, !empty($strategies));
```
Comment (Ener-Getick) [inline Tests/Context/ContextTest.php]: The next assertion is enough

File: Tests/Context/ContextTest.php [hunk 7]
```diff
@@ -98,4 +98,20 @@ public function testSerializeNull()
...
+        $this->assertEquals(true, in_array($strategy1, $strategies, true));
+        $this->assertEquals(true, in_array($strategy2, $strategies, true));
```
Comment (Ener-Getick) [inline Tests/Context/ContextTest.php]: Can be replaced by assertEquals([$s1, $s2], $strategies);

**Comment by bblue** [comment]
Done. That was close to my first ever test, so don't ask me to create another unless you are prepared to tear it apart :)

File: Tests/Context/ContextTest.php [hunk 8]
```diff
@@ -98,4 +98,17 @@ public function testSerializeNull()
 
         $this->assertEquals(true, $this->context->getSerializeNull());
     }
+
+    public function testExclusionStrategy()
+    {
+        $strategy1 = $this->getMock('JMS\Serializer\Exclusion\ExclusionStrategyInterface');
+        $strategy2 = $this->getMock('JMS\Serializer\Exclusion\ExclusionStrategyInterface');
```
Comment (Ener-Getick) [inline Tests/Context/ContextTest.php]: This is deprecated :/ you can replace it by `getMockBuilder(...)->getMock()` and also use the `::class` constant.

**Comment by Ener-Getick** [comment]
@bblue no problem, my review is here for that :)
Thanks for taking care of this issue !

**Comment by Ener-Getick** [comment]
Thank you @bblue !

**Comment by xabbuh** [comment]
The `Context` class is part of our abstraction layer. Cluttering it with implementation-specific (JMS-specific here) code goes against this abstraction. I think a proper solution would be to ship our own `ExclusionStrategy` interface and let the concrete adapters deal with the implementation-specific stuff.

This way we could also easily update the Symfony Serializer integration if at some time in the future the Symfony Serializer component ships with a similar feature.

**Comment by Ener-Getick** [comment]
@xabbuh i'm affraid it would be confusing and harder to use for no gain.
What's the problem of depending on an external interface as long as we can still use the class in other contexts? At worst if we really don't want a jms interface what about using an attribute?

**Comment by xabbuh** [comment]
Imagine the Symfony Serializer got exclusion strategies in the future too? Due to the JMS type hint we would have to add new methods for the Symfony implementation serving the same purpose but with a different signature. Or what do we do when we want to support more implementations (for example, [ivory/serializer](https://github.com/egeloen/ivory-serializer/)? This would lead in the same direction. If we went this way, the `Context` class could become a huge mess.

**Comment by Ener-Getick** [comment]
@xabbuh we can update the class signature (it is final) so we would only have to replace the type hint by a less stricter one.
BTW what do you think about recentrating the Context class around the attributes concept? It looks weird to me that `setAttribute('groups')` does not the same as `setGroups()`.

**Comment by xabbuh** [comment]
But what would the less stricter type hint? There is no common base interface that we could use. Which means that we would then be forced to remove the type hint at all.

What exactly do you mean with the attributes concept? I mean what do you envision?

**Comment by Ener-Getick** [comment]
If symfony adds exclusion strategies support, we could add its interface to our type hinting, but i don't think it's worth having a common class.

For the attributes, i just would like something like:
```php
public function getGroups() {
    return $this->getAttribute('groups');
}
```
That would be a small bc break, but the class behavior would be more expectable.

**Comment by xabbuh** [comment]
We couldn't add the interface to the type hint (except if there was a common base interface for both implementations which doesn't seem very likely). So effectively we would have to remove the type hint. I am not convinced if that were the better option. At least it would make it harder to find some bugs in userland code.

Speaking about the attributes: Making this change would be compatible with how the Symfony serializer handles groups and attributes. So it would make the behaviour more consistent with this implementation (you couldn't have a custom `groups` attribute here anyway). However, we would introduce a small BC break for the JMS serializer like you said. I am actually not completely sure if we should go that way (it may confuse people who are used to the JMS serializer that there are some attributes that have a special meaning).

**Comment by GuilhemN** [comment]
Sorry i forgot to answer you...
I don't think the symfony serializer will ever support exclusion strategies so imo it's fine to couple this feature to the jms serializer.

**Pull request merged**